link/tracing.go

```diff
@@ -87,29 +88,72 @@ type TracingOptions struct {
 	// AttachTraceFEntry/AttachTraceFExit/AttachModifyReturn or
 	// AttachTraceRawTp.
 	Program *ebpf.Program
+	// Program attach type. Can be one of:
+	// 	- AttachTraceFEntry
+	// 	- AttachTraceFExit
+	// 	- AttachModifyReturn
+	// 	- AttachTraceRawTp
+	// If empty (AttachNone), the legacy attach method via raw tracepoint
+	// open will be used.
+	AttachType ebpf.AttachType
+	// Arbitrary value that can be fetched from an eBPF program
+	// via `bpf_get_attach_cookie()`.
+	Cookie uint64
 }
 
 type LSMOptions struct {
 	// Program must be of type LSM with attach type
 	// AttachLSMMac.
 	Program *ebpf.Program
+	// Arbitrary value that can be fetched from an eBPF program
+	// via `bpf_get_attach_cookie()`.
+	Cookie uint64
 }
 
 // attachBTFID links all BPF program types (Tracing/LSM) that they attach to a btf_id.
-func attachBTFID(program *ebpf.Program) (Link, error) {
+func attachBTFID(program *ebpf.Program, at ebpf.AttachType, cookie uint64) (Link, error) {
 	if program.FD() < 0 {
 		return nil, fmt.Errorf("invalid program %w", sys.ErrClosedFd)
 	}
 
-	fd, err := sys.RawTracepointOpen(&sys.RawTracepointOpenAttr{
-		ProgFd: uint32(program.FD()),
-	})
-	if errors.Is(err, sys.ENOTSUPP) {
-		// This may be returned by bpf_tracing_prog_attach via bpf_arch_text_poke.
-		return nil, fmt.Errorf("create raw tracepoint: %w", ErrNotSupported)
-	}
-	if err != nil {
-		return nil, fmt.Errorf("create raw tracepoint: %w", err)
+	var (
+		fd  *sys.FD
+		err error
+	)
+	switch at {
+	case ebpf.AttachTraceFEntry, ebpf.AttachTraceFExit, ebpf.AttachTraceRawTp,
+		ebpf.AttachModifyReturn, ebpf.AttachLSMMac:
+		// Attach via BPF link
+		fd, err = sys.LinkCreateTracing(&sys.LinkCreateTracingAttr{
+			ProgFd:     uint32(program.FD()),
+			AttachType: sys.AttachType(at),
+			Cookie:     cookie,
+		})
+		if err == nil {
+			break
+		}
+		if !errors.Is(err, unix.EINVAL) {
+			return nil, fmt.Errorf("create tracing link: %w", err)
+		}
+		fallthrough
+	case ebpf.AttachNone:
```

Comment:
What is the reason the default attach type can't use bpf_link?

Reply:
`AttachType` needs to be specified in `LinkCreateTracingAttr`

Reply:
https://elixir.bootlin.com/linux/v6.2.2/source/kernel/bpf/syscall.c#L4577
```C
		ret = bpf_tracing_prog_attach(prog,
					      attr->link_create.target_fd,
					      attr->link_create.target_btf_id,
					      attr->link_create.tracing.cookie);
```

https://elixir.bootlin.com/linux/v6.2.2/source/kernel/bpf/syscall.c#L4595
```C
			ret = bpf_tracing_prog_attach(prog,
						      attr->link_create.target_fd,
						      attr->link_create.target_btf_id,
						      attr->link_create.tracing.cookie);
```

https://elixir.bootlin.com/linux/v6.2.2/source/kernel/bpf/syscall.c#L3310
```C
		return bpf_tracing_prog_attach(prog, 0, 0, 0);
```

* Both RAW_TRACEPOINT_OPEN and BPF_LINK_CREATE return a link fd (on newer kernels)
* But RAW_TRACEPOINT_OPEN doesn't allow specifying target or cookie

On new kernels, RAW_TRACEPOINT_OPEN == BPF_LINK_CREATE if target_fd, cookie, etc. are 0.

Thinking through the compatibility story here: seems like specifying AttachType on a kernel without LINK_CREATE will fail. So it's not possible to just always set AttachType if you want maximum compat, instead you should only set it if `Cookie != 0`. Can we do better?

Reply:
So, to recap:

- on new kernels it's always better to use `LINK_CREATE` as it behaves the same way if cookie etc. are 0
- on old kernels we can fail and fallback (fallthrough) to `RAW_TRACEPOINT_OPEN`

> seems like specifying AttachType on a kernel without LINK_CREATE will fail

this should be handled by `fallthrough`, right?

> So it's not possible to just always set AttachType if you want maximum compat, instead you should only set it if Cookie != 0

could you clarify this point? If I set `opts.AttachType` on a kernel that doesn't support `LINK_CREATE`, I expect it to fail and fallback to `RAW_TRACEPOINT_OPEN`

Reply:
> this should be handled by `fallthrough`, right?

yes, my bad, sorry!

> * on new kernels it's always better to use `LINK_CREATE` as it behaves the same way if cookie etc. are 0

Nit: seems to me like LINK_CREATE vs RAW_TRACEPOINT_OPEN makes no difference if cookie is 0. They end up calling the same function?

Agreed that we should prefer LINK_CREATE.